Networking/Networking/ProductImageInBackground/MediaUploadSessionManager.swift

```diff
@@ -0,0 +1,164 @@
+import Foundation
+
+public protocol MediaUploadSessionManagerDelegate: AnyObject {
+    func mediaUploadSessionManager(_ manager: MediaUploadSessionManager,
+                                   didCompleteUpload uploadID: String,
+                                   withResult result: Result<Media, Error>)
+}
+
+/// Background upload specific errors
+enum BackgroundUploadError: Error {
+    case invalidRequestBody
+    case invalidResponse
+    case decodingError
+}
+
+/// Session Manager for media upload in background
+///
+public final class MediaUploadSessionManager: NSObject {
+
+    public let backgroundSessionIdentifier: String
+    private lazy var backgroundSession: URLSession = {
+        let config = URLSessionConfiguration.background(withIdentifier: backgroundSessionIdentifier)
+        config.sharedContainerIdentifier = "group.com.automattic.woocommerce"
+        config.sessionSendsLaunchEvents = true
+        config.isDiscretionary = false
+        config.allowsCellularAccess = true
+        return URLSession(configuration: config, delegate: self, delegateQueue: nil)
+    }()
+
+    private var completionHandlers: [String: (Result<Media, Error>) -> Void] = [:]
+    private var taskResponseData: [Int: Data] = [:]
+    private var backgroundCompletionHandler: (() -> Void)?
+    weak var delegate: MediaUploadSessionManagerDelegate?
+
+    public init(sessionIdentifier: String = "com.automattic.woocommerce.background.upload") {
+        self.backgroundSessionIdentifier = sessionIdentifier
+        super.init()
+    }
+
+    public func uploadMedia(request: URLRequest,
+                            mediaItem: UploadableMedia,
+                            uploadID: String,
+                            completion: @escaping (Result<Media, Error>) -> Void) {
+        completionHandlers[uploadID] = completion
+
+        guard let httpBody = request.httpBody else {
+            let error = BackgroundUploadError.invalidRequestBody
+            completion(.failure(error))
+            return
+        }
+
+        do {
+            // Create temp file with proper extension from mediaItem
+            let tempDirectory = FileManager.default.temporaryDirectory
+            let tempFileURL = tempDirectory.appendingPathComponent(mediaItem.filename)
+            try httpBody.write(to: tempFileURL)
+
+            // Upload using temp file
+            var modifiedRequest = request
+            modifiedRequest.httpBody = nil
+
+            let task = backgroundSession.uploadTask(with: modifiedRequest, fromFile: tempFileURL)
```

Comment:
I see that we have already appended the data of the media item to the request body in [this line](https://github.com/woocommerce/woocommerce-ios/pull/15277/files#diff-11e05f6bd757dd53ad88ab050fb76c0104ae789632f6ba30f581286fd65bc671R226). Do we still have to create a temporary file and pass it to this `uploadTask` method? 

Is [this the reason](https://www.avanderlee.com/swift/urlsession-common-pitfalls-with-background-download-upload-tasks/#3-only-urlsession-upload-tasks-from-a-file-support-background-execution) for using a temp file? If yes, can you add a comment to explain?

Reply:
That's exactly the reason. I will add a comment to clarify that for the future 👍